### PR TITLE
4.next - Explicit list of supported valueSources

### PR DESCRIPTION
**Pull request by diegosardina**
No functional change, just removed hardcoded part. This will reduce code of my next PR.

Tests already cover for valid values sources

### CODE REVIEW

File: src/View/Helper/FormHelper.php [hunk 1]
```diff
@@ -2480,18 +2487,29 @@ public function getValueSources(): array
         return $this->_valueSources;
     }
 
+    /**
+     * Extracts valid value sources.
+     *
+     * @param string|string[] $sources A string or a list of strings identifying a source.
+     * @return array
```
Comment (dereuromark) [inline src/View/Helper/FormHelper.php:2494]: string[] then?

Comment (diegosardina) [reply src/View/Helper/FormHelper.php:2494]: It receives the same type of setValueSources()

Comment (dereuromark) [reply src/View/Helper/FormHelper.php:2494]: I don't think the return result does, if you look into what the code does.

File: src/View/Helper/FormHelper.php [hunk 2]
```diff
@@ -2480,18 +2487,29 @@ public function getValueSources(): array
...
+     * Valid values are in $this->_supportedValueSources.
      * You need to supply one valid context or multiple, as a list of strings. Order sets priority.
      *
      * @param string|string[] $sources A string or a list of strings identifying a source.
      * @return $this
      */
     public function setValueSources($sources)
     {
-        $this->_valueSources = array_values(array_intersect((array)$sources, ['context', 'data', 'query']));
+        $this->_valueSources = $this->_extractValidValueSources($sources);
```
Comment (garas) [inline src/View/Helper/FormHelper.php:2512]: Use `(array)` here, and require `array` as argument innew method.

File: src/View/Helper/FormHelper.php [hunk 3]
```diff
@@ -2480,18 +2487,29 @@ public function getValueSources(): array
         return $this->_valueSources;
     }
 
+    /**
+     * Extracts valid value sources.
+     *
+     * @param string[] $sources A a list of strings identifying a source.
```
Comment (garas) [inline src/View/Helper/FormHelper.php:2493]:
```suggestion
     * @param string[] $sources A list of strings identifying a source.
```

File: src/View/Helper/FormHelper.php [hunk 4]
```diff
@@ -2480,18 +2487,29 @@ public function getValueSources(): array
         return $this->_valueSources;
     }
 
+    /**
+     * Extracts valid value sources.
+     *
+     * @param string[] $sources A list of strings identifying a source.
+     * @return string[]
+     */
+    protected function _extractValidValueSources(array $sources): array
```
Comment (othercorey) [inline src/View/Helper/FormHelper.php:2496]:
```suggestion
    protected function extractValidValueSources(array $sources): array
```

File: src/View/Helper/FormHelper.php [hunk 5]
```diff
@@ -2480,18 +2487,29 @@ public function getValueSources(): array
...
+    {
+        return array_values(array_intersect($sources, $this->_supportedValueSources));
+    }
+
     /**
      * Sets the value sources.
      *
-     * Valid values are `'context'`, `'data'` and `'query'`.
+     * Valid values are in $this->_supportedValueSources.
```
Comment (othercorey) [inline src/View/Helper/FormHelper.php:2504]:
```suggestion
     * Valid values are returned by `supportedValueSources()`.
```

File: src/View/Helper/FormHelper.php [hunk 6]
```diff
@@ -2480,18 +2487,29 @@ public function getValueSources(): array
...
      * You need to supply one valid context or multiple, as a list of strings. Order sets priority.
      *
      * @param string|string[] $sources A string or a list of strings identifying a source.
      * @return $this
      */
     public function setValueSources($sources)
     {
-        $this->_valueSources = array_values(array_intersect((array)$sources, ['context', 'data', 'query']));
+        $this->_valueSources = $this->_extractValidValueSources((array)$sources);
```
Comment (othercorey) [inline src/View/Helper/FormHelper.php:2512]:
```suggestion
        $this->_valueSources = $this->extractValidValueSources((array)$sources);
```

File: src/View/Helper/FormHelper.php [hunk 7]
```diff
@@ -219,7 +219,14 @@ class FormHelper extends Helper
     protected $_lastAction = '';
 
     /**
-     * The sources to be used when retrieving prefilled input values.
+     * The supported sources that can be used to populate input values.
+     *
+     * @var string[]
+     */
+    protected $_supportedValueSources = ['context', 'data', 'query'];
```
Comment (othercorey) [inline src/View/Helper/FormHelper.php:226]:
```suggestion
    protected $supportedValueSources = ['context', 'data', 'query'];
```

File: src/View/Helper/FormHelper.php [hunk 8]
```diff
@@ -2480,18 +2487,29 @@ public function getValueSources(): array
...
 
+    /**
+     * Extracts valid value sources.
+     *
+     * @param string[] $sources A list of strings identifying a source.
+     * @return string[]
+     */
+    protected function _extractValidValueSources(array $sources): array
+    {
+        return array_values(array_intersect($sources, $this->_supportedValueSources));
```
Comment (othercorey) [inline src/View/Helper/FormHelper.php:2498]:
```suggestion
        return array_values(array_intersect($sources, $this->supportedValueSources));
```

**Comment by markstory** [comment]
Seems fine, I don't understand why we need the new method yet though.

**Comment by dereuromark** [comment]
Extendability probably, if you want to add in another source.

**Comment by diegosardina** [comment]
> Seems fine, I don't understand why we need the new method yet though.

It is for my next PR. I found a case where it's convenient to override the valueSources just for a single field. The code is very small and doesn't add complexity to the FormHelper and imho it is worth to have.

However I didn't like to have this in _two places_:

```
... = array_values(array_intersect((array)$sources, ['context', 'data', 'query']));
```

Another reason to isolate that part is to raise an exception in the case the list will be empty (for example if I mistakenly wrote ['adta'] instead of ['data'], the field will never get a value). But I didn't because it's a behavioural change (it also requires tests adaptation) and I wanted this to be less invasive as possible in a minor release.

**Comment by diegosardina** [comment]
> Extendability probably, if you want to add in another source.

Yup, maybe one day 'session' could be added, but I don't need it now :-)

**Pull request merged**